File: fotoapparat-adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/SingleAdapter.java
package io.fotoapparat.result.adapter.rxjava;

import java.util.concurrent.Future;

import io.fotoapparat.result.adapter.Adapter;
import rx.Single;

/**
 * Adapter for {@link Single}.
 */
public class SingleAdapter<T> implements Adapter<T, Single<T>> {

    private SingleAdapter() {
    }

    /**
     * @return {@link Adapter} which adapts result to {@link Single}.
     */
    public static <R> SingleAdapter<R> toSingle() {
        return new SingleAdapter<>();
    }

    @Override
    public Single<T> adapt(Future<T> future) {
        return Single.from(future);
    }
}

File: fotoapparat-adapters/rxjava/src/main/java/io/fotoapparat/result/adapter/rxjava/SingleAdapter.kt
package io.fotoapparat.result.adapter.rxjava

import io.fotoapparat.result.PendingResult
import rx.Single
import java.util.concurrent.Future

/**
 * Adapter for [Single].
 */
object SingleAdapter {

    /**
     * @return Adapter which adapts result to [Single].
     */
    @JvmStatic
    fun <T> toSingle(): (Future<T>) -> Single<T> {
        return { future -> Single.from(future) }
    }

}

/**
 * @return A [Single] from the given [PendingResult].
 */
fun <T> PendingResult<T>.toSingle(): Single<T> {
    return adapt { future -> Single.from(future) }
}
